packages/zoe/src/contractFacet/contractFacet.js

```diff
@@ -0,0 +1,196 @@
+        // Discussion at: https://github.com/Agoric/agoric-sdk/issues/1017
+        assert(
+          seatStagings.length >= 2,
+          details`reallocating must be done over two or more seats`,
+        );
+
+        seatStagings.forEach(seatStaging =>
+          assert(
+            allSeatStagings.has(seatStaging),
+            details`The seatStaging ${seatStaging} was not recognized`,
+          ),
+        );
+
+        // Ensure that rights are conserved overall. Offer safety was
+        // already checked when an allocation was staged for an individual seat.
+        const flattened = arr => [].concat(...arr);
+        const flattenAllocations = allocations =>
+          flattened(allocations.map(allocation => Object.values(allocation)));
+
+        const previousAllocations = seatStagings.map(seatStaging =>
+          seatStaging.getSeat().getCurrentAllocation(),
+        );
+        const previousAmounts = flattenAllocations(previousAllocations);
+
+        const newAllocations = seatStagings.map(seatStaging =>
+          seatStaging.getStagedAllocation(),
+        );
+        const newAmounts = flattenAllocations(newAllocations);
+
+        assert(
+          areRightsConserved(getAmountMath, previousAmounts, newAmounts),
+          details`Rights are not conserved in the proposed reallocation`,
+        );
+
+        // Commit the staged allocations and inform Zoe of the
+        // newAllocation.
+        seatStagings.forEach(seatStaging =>
+          seatToSeatAdmin.get(seatStaging.getSeat()).commit(seatStaging),
+        );
+      },
+      saveIssuer: (issuerP, keyword) => {
+        return E(zoeInstanceAdmin)
+          .saveIssuer(issuerP, keyword)
+          .then(() => {
+            return issuerTable
+              .getPromiseForIssuerRecord(issuerP)
+              .then(issuerRecord => {
+                assertKeywordName(keyword);
+                assert(
+                  !getKeywords(instanceRecord.issuerKeywordRecord).includes(
+                    keyword,
+                  ),
+                  details`keyword ${keyword} must be unique`,
+                );
+                instanceRecord.issuerKeywordRecord = {
+                  ...instanceRecord.issuerKeywordRecord,
+                  [keyword]: issuerRecord.issuer,
+                };
+                instanceRecord.brandKeywordRecord = {
+                  ...instanceRecord.brandKeywordRecord,
+                  [keyword]: issuerRecord.brand,
+                };
+
+                return issuerRecord;
+              });
+          });
+      },
+      makeInvitation: (offerHandler, description, customProperties = {}) => {
+        assert.typeof(
+          description,
+          'string',
+          details`invitations must have a description string: ${description}`,
+        );
+        /** @type {InvitationHandle} */
+        const invitationHandle = {};
+        harden(invitationHandle);
```

Comment:
Why not say
```js
const invitationHandle = harden({});
```
? If it is because of `harden`'s typing, I think @michaelfig fixed that.

Reply:
The typing wouldn't be recognized without this. Unless it was something that was fixed in the past week or so (this PR needs to be rebased), I think it's still an issue.

Reply:
@michaelfig is this typing of `harden` fixed?